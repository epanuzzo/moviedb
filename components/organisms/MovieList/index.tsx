import React from "react";
import MovieCard from "@/components/molecules/MovieCard";
import { useSelector } from "react-redux";
import { RootState } from "@/redux/store";
import { imageUrl } from "@/utils/config";

/**
 * MovieList component displays a list of movies fetched from the Redux store.
 *
 * This component uses the `useSelector` hook to access the `movies` and `error`
 * properties from the `movie` slice of the Redux state.
 *
 * If there is an error, it displays the error message.
 * If there are no movies, it displays a "No movies found!" message.
 * Otherwise, it renders a grid of `MovieCard` components, each representing a movie.
 *
 * @component
 * @returns {JSX.Element} The rendered component.
 */
const MovieList: React.FC = () => {
  const { movies, error } = useSelector((state: RootState) => state.movie);

  if (error) {
    return <h1 className="text-4xl font-bold text-center">{error}</h1>;
  }

  if (!movies.length) {
    return <h1 className="text-4xl font-bold text-center">No movies found!</h1>;
  }

  return (
    <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4 p-4">
      {movies.map((movie) => (
        <MovieCard
          id={movie.id}
          title={movie.title}
          rating={movie.vote_average}
          image={imageUrl + movie.poster_path}
          missingImage={!movie.poster_path}
          key={movie.title}
        />
      ))}
    </div>
  );
};

export default MovieList;
