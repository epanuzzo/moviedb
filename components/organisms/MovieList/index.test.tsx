import React from "react";
import renderer from "react-test-renderer";
import MovieList from "./index";
import { useSelector } from "react-redux";

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
}));

describe("MovieList", () => {
  beforeEach(() => {
    (useSelector as unknown as jest.Mock).mockReturnValue({
      movies: [
        { id: 1, title: "Movie 1" },
        { id: 2, title: "Movie 2" },
      ],
      loading: false,
      pages: 2,
      currentPage: 1,
    });
  });

  it("renders the component", () => {
    const tree = renderer.create(<MovieList />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with error", () => {
    (useSelector as unknown as jest.Mock).mockReturnValueOnce({
      movies: [],
      loading: true,
      pages: 0,
      currentPage: 0,
      error: "error",
    });

    const tree = renderer.create(<MovieList />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with no movies", () => {
    (useSelector as unknown as jest.Mock).mockReturnValueOnce({
      movies: [],
      loading: true,
      pages: 0,
      currentPage: 0,
    });

    const tree = renderer.create(<MovieList />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
