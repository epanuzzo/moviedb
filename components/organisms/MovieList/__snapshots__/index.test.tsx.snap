// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList renders the component 1`] = `
<div
  className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4 p-4"
>
  <a
    className="relative bg-white rounded-md shadow-lg overflow-hidden h-80 hover:cursor-pointer"
    href="/movie/1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="w-full h-full bg-gray-800 flex items-center justify-center"
    >
      <p
        className="text-white"
      >
        Image not found
      </p>
    </div>
    <div
      className="absolute bottom-0 left-0 right-0 p-4 bg-gradient-to-t from-black to-transparent"
    >
      <h3
        className="text-lg font-semibold text-white"
      >
        Movie 1
      </h3>
    </div>
    <div
      className="absolute top-4 left-4 bg-white text-black rounded-full flex items-center justify-center w-10 h-10"
    >
      <p
        className="text-sm font-bold"
      />
    </div>
  </a>
  <a
    className="relative bg-white rounded-md shadow-lg overflow-hidden h-80 hover:cursor-pointer"
    href="/movie/2"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="w-full h-full bg-gray-800 flex items-center justify-center"
    >
      <p
        className="text-white"
      >
        Image not found
      </p>
    </div>
    <div
      className="absolute bottom-0 left-0 right-0 p-4 bg-gradient-to-t from-black to-transparent"
    >
      <h3
        className="text-lg font-semibold text-white"
      >
        Movie 2
      </h3>
    </div>
    <div
      className="absolute top-4 left-4 bg-white text-black rounded-full flex items-center justify-center w-10 h-10"
    >
      <p
        className="text-sm font-bold"
      />
    </div>
  </a>
</div>
`;

exports[`MovieList renders with error 1`] = `
<h1
  className="text-4xl font-bold text-center"
>
  error
</h1>
`;

exports[`MovieList renders with no movies 1`] = `
<h1
  className="text-4xl font-bold text-center"
>
  No movies found!
</h1>
`;
